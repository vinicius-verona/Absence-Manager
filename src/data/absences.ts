import _absences from "./absences.json";

export type Absence = {
    admitterId: number | null;
    admitterNote: string;
    confirmedAt: Date | null;
    createdAt: string;
    crewId: number;
    endDate: string;
    id: number;
    memberNote: string;
    rejectedAt: Date | null;
    startDate: string;
    type: string;
    userId: number;
    status: string;
};

/**
 * Each absence from `absences.json` is inserted into a map.
 *
 * @returns Map<number, Absence> with absence's `id` as key.
 */
export function getAbsences() {
    const absences = new Map<number, Absence>();

    _absences.payload.forEach((_absence) => {
        let _content = _absence as Absence;

        if (!_absence.confirmedAt && !_absence.rejectedAt) {
            _content.status = "Requested";
        } else if (_absence.confirmedAt) {
            _content.status = "Confirmed";
        } else {
            _content.status = "Rejected";
        }

        absences.set(_content.id, _content);
    });

    return absences;
}

export const ABSENCES = getAbsences();
