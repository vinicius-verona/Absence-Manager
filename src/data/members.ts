import _members from "./members.json";

type Member = {
    crewId: number;
    id: number;
    image: string;
    name: string;
    userId: number;
};

/**
 * Each member from `members.json` is inserted into a map.
 *
 * @returns Map<number, Member> with member's `userId` as key.
 */
export function getMembers() {
    const members = new Map<number, Member>();

    _members.payload.forEach((_member) => {
        members.set(_member.userId, _member);
    });

    return members;
}
