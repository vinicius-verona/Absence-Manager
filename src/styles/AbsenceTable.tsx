import styled from "styled-components";

export const Table = styled.table`
    width: 100%;
    align-items: center;
    justify-content: center;
    padding: 10px 0px;
    text-align: center;

    tr:hover td {
        background-color: ${({ theme }) => theme.palettes.lightPrimary};
    }
    .dark-row {
        background-color: ${({ theme }) => theme.palettes.light};
    }
`;
