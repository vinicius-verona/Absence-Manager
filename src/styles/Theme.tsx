// eslint-disable-next-line import/no-anonymous-default-export
export default {
    palettes: {
        primary: "rgba(255, 148, 25, 1)",
        lightPrimary: "rgba(255, 148, 25, 0.5)",
        light: "#DFDFDF",
        darkerLight: "#B5B5B5",
        bg_dark: "rgba(0, 0, 0, 0.5)",
    },
};
