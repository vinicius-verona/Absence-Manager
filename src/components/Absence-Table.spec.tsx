import { render, waitFor } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { Absence } from "../data/absences";
import { Member, MEMBERS } from "../data/members";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import Theme from "../styles/Theme";
import AbsenceTable from "./Absence-Table";

describe("Render absence table on page", () => {
  const absence = {
    admitterId: null,
    admitterNote: "",
    confirmedAt: "2020-12-12T18:03:55.000+01:00",
    createdAt: "2020-12-12T14:17:01.000+01:00",
    crewId: 352,
    endDate: "2021-01-13",
    id: 2351,
    memberNote: "",
    rejectedAt: null,
    startDate: "2021-01-13",
    type: "sickness",
    userId: 2664,
    status: "Confirmed",
  };

  const absences = new Map<number, Absence>();
  absences.set(absence.id, absence);

  const member = MEMBERS.get(absence.userId);
  if (!member) {
    throw "Unexisting member";
  }

  const members = new Map<number, Member>();
  members.set(member.userId, member);

  if (!member) {
    throw "Member not found";
  }

  it("Should match absence table snapshot", () => {
    const table = renderer
      .create(
        <ThemeProvider theme={Theme}>
          <AbsenceTable absences={absences} members={members} />
        </ThemeProvider>
      )
      .toJSON();

    expect(table).toMatchSnapshot();
  });

  it("Should render table", () => {
    const { getByTestId } = render(
      <ThemeProvider theme={Theme}>
        <AbsenceTable absences={absences} members={members} />
      </ThemeProvider>
    );

    expect(getByTestId("absence-table")).toBeInTheDocument();
  });

  it("Should render absence row", () => {
    const { getByText } = render(
      <ThemeProvider theme={Theme}>
        <AbsenceTable absences={absences} members={members} />
      </ThemeProvider>
    );

    expect(getByText(absence.id)).toBeInTheDocument();
    expect(getByText(member.name)).toBeInTheDocument();
    expect(getByText(member.crewId)).toBeInTheDocument();
    expect(getByText(absence.type)).toBeInTheDocument();
    expect(getByText(absence.status)).toBeInTheDocument();
  });

  it("Should render modal for selected absence", async () => {
    const { getByText } = render(
      <ThemeProvider theme={Theme}>
        <AbsenceTable absences={absences} members={members} />
      </ThemeProvider>
    );

    expect(getByText(absence.id)).toBeInTheDocument();
    userEvent.click(getByText(absence.id));
    waitFor(() => {
      expect(getByText(`Absence ${absence.id}`)).toBeInTheDocument();
    });
  });
});
