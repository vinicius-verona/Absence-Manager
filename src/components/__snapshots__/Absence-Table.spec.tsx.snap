// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render absence table on page Should match absence table snapshot 1`] = `
<table
  className="sc-bdvvtL ecXJRw"
  data-testid="absence-table"
>
  <thead>
    <tr>
      <th>
        Absence
      </th>
      <th>
        Member
      </th>
      <th>
        Crew
      </th>
      <th>
        Type
      </th>
      <th>
        Status
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className=""
      onClick={[Function]}
    >
      <td>
        2351
      </td>
      <td>
        Mike
      </td>
      <td>
        352
      </td>
      <td>
        sickness
      </td>
      <td>
        Confirmed
      </td>
    </tr>
  </tbody>
</table>
`;
